### PR TITLE
Add new keyword code fix

### PR DESCRIPTION
**Pull request by vasily-kirichenko**
image

### CODE REVIEW

**Comment by forki** [comment]
Cool,should it also change let to use?

File: vsintegration/src/FSharp.Editor/CodeFix/AddOpenCodeFixProvider.fs [hunk 1]
```diff
@@ -108,6 +108,7 @@ type internal FSharpAddOpenCodeFixProvider
         assemblyContentProvider: AssemblyContentProvider
     ) =
     inherit CodeFixProvider()
+    let fixableDiagnosticIds = ["FS0039"]
```
Comment (forki) [inline vsintegration/src/FSharp.Editor/CodeFix/AddOpenCodeFixProvider.fs]: Could you please apply the same pattern to the other fix provider?

Comment (vasily-kirichenko) [inline vsintegration/src/FSharp.Editor/CodeFix/AddOpenCodeFixProvider.fs]: what do you mean? Introduce an extra field for diagnostic codes? why?

Comment (forki) [inline vsintegration/src/FSharp.Editor/CodeFix/AddOpenCodeFixProvider.fs]: I think both should follow similar pattern.

**Comment by vasily-kirichenko** [comment]
> Cool,should it also change let to use?

No. I think it was discussed somewhere on github, in short it's not always feasible to use `use` instead of `let` because it's completely legible to create a disposable object, but not bind it with `use`.

File: vsintegration/src/FSharp.Editor/CodeFix/AddOpenCodeFixProvider.fs (hunk 1 above)
Comment (vasily-kirichenko) [inline vsintegration/src/FSharp.Editor/CodeFix/AddOpenCodeFixProvider.fs]: It's maybe better to introduce a base class, but, you know...

**Comment by forki** [comment]
Yeah, but if we suggest with reasonable explanation then it can still be useful. And people can ignore suggestions, right?

File: vsintegration/src/FSharp.Editor/CodeFix/AddOpenCodeFixProvider.fs (hunk 1 above)
Comment (forki) [inline vsintegration/src/FSharp.Editor/CodeFix/AddOpenCodeFixProvider.fs]: I didn't say that ;-) 
Just saying if we find a common pattern than this is beneficial.

Comment (nosami) [inline vsintegration/src/FSharp.Editor/CodeFix/AddOpenCodeFixProvider.fs]: This acts upon `FS0760`

File: vsintegration/src/FSharp.Editor/CodeFix/AddNewKeywordToDisposableConstructorInvokation.fs [hunk 2]
```diff
@@ -0,0 +1,33 @@
+// Copyright (c) Microsoft Corporation.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
```
Comment (nosami) [inline vsintegration/src/FSharp.Editor/CodeFix/AddNewKeywordToDisposableConstructorInvokation.fs]: typo in Filename... should be `AddNewKeywordToDisposableConstructorInvocation.fs`

Comment (vasily-kirichenko) [inline vsintegration/src/FSharp.Editor/CodeFix/AddNewKeywordToDisposableConstructorInvokation.fs]: Thanks ;)

**Comment by vasily-kirichenko** [comment]
> Yeah, but if we suggest with reasonable explanation then it can still be useful. And people can ignore suggestions, right?

This require writing an Analyzer which would emit special warning, then a Code Fix should be written, which suggests possible fixes. This PR is so simple because the compiler emits a warning itself.

**Comment by forki** [comment]
Another point for making the compiler smart ;-)

Am 22.12.2016 2:29 nachm. schrieb "Vasily Kirichenko" <
notifications@github.com>:

**Comment by cartermp** [comment]
This is fantastic.  With respect to `use`, I think it's one of those areas of the language where we'll have to make a decision on what we want people to ultimately be using.  Both usages are valid, but if we agree that there is a preferred way that the IDE should _drive_ you towards, we can modify this to suggest that way.

File: vsintegration/src/FSharp.Editor/CodeFix/AddNewKeywordToDisposableConstructorInvocation.fs [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+
+[<ExportCodeFixProvider(FSharpCommonConstants.FSharpLanguageName, Name = "AddNewKeyword"); Shared>]
+type internal FSharpAddNewKeywordCodeFixProvider() =
+    inherit CodeFixProvider()
+
+    override __.FixableDiagnosticIds = ["FS0760"].ToImmutableArray()
+
+    override this.RegisterCodeFixesAsync context : Task =
+        async {
+            let title = "Add \"new\" keyword"
```
Comment (KevinRansom) [inline vsintegration/src/FSharp.Editor/CodeFix/AddNewKeywordToDisposableConstructorInvocation.fs]: Localizable text ...

**Comment by smoothdeveloper** [comment]
@cartermp, both use and let are valid and have different semantics so I don't think there is a way to have a single selection.

I agree having both proposed (let being first) would make sense.

**Comment by cartermp** [comment]
@smoothdeveloper Right, but this is why I think it's worth having a discussion about how to approach code fixes in general.  There are lots of cases where there are multiple valid ways to do something.  In some cases it's best to show all options, and in other cases it may be best to show the preferred option.  This is something we'll just have to evaluate on a case-by-case basis.

File: vsintegration/src/FSharp.Editor/CodeFix/AddNewKeywordToDisposableConstructorInvocation.fs (hunk 3 above)
Comment (vasily-kirichenko) [inline vsintegration/src/FSharp.Editor/CodeFix/AddNewKeywordToDisposableConstructorInvocation.fs]: Yes. Could you explain how to do it? (maybe it's worth to add somewhere into DEVGUIDE.md?)

Comment (KevinRansom) [inline vsintegration/src/FSharp.Editor/CodeFix/AddNewKeywordToDisposableConstructorInvocation.fs]: @vasily-kirichenko 
I will submit a PR with the fix and then we can have a document that points to commit with the changes ... that seems like the most straightforward approach.

Kevin

**Pull request merged**